miner/worker.go

```diff
@@ -288,12 +288,25 @@ func (w *worker) start() {
 	w.startCh <- struct{}{}
 
 	if istanbul, ok := w.engine.(consensus.Istanbul); ok {
-		istanbul.SetBlockProcessors(w.chain.HasBadBlock,
+		istanbul.SetCallBacks(w.chain.HasBadBlock,
 			func(block *types.Block, state *state.StateDB) (types.Receipts, []*types.Log, uint64, error) {
 				return w.chain.Processor().Process(block, state, *w.chain.GetVMConfig())
 			},
-			func(block *types.Block, state *state.StateDB, receipts types.Receipts, usedGas uint64) error {
-				return w.chain.Validator().ValidateState(block, state, receipts, usedGas)
+			w.chain.Validator().ValidateState,
+			func(block *types.Block, receipts []*types.Receipt, logs []*types.Log, state *state.StateDB) {
+				if err := w.chain.InsertPreprocessedBlock(block, receipts, logs, state); err != nil {
+					if err == core.ErrNotHeadBlock {
+						log.Warn("Tried to insert duplicated produced block", "blockNumber", block.Number(), "hash", block.Hash(), "err", err)
+					} else {
+						log.Error("Failed to insert produced block", "blockNumber", block.Number(), "hash", block.Hash(), "err", err)
```

Comment:
Its not super clear to me the meaning of all the errors that could be returned here, but I would think some of those errors indicate a consensus breakdown or programming error and for those errors we should stop processing.

For example if the randomness cannot be generated, that looks like it is a serious failure, not one we would want to keep processing on top of.

Reply:
https://github.com/celo-org/celo-blockchain/blob/68e2ab747b40ce215a916cde7fdc77613366dcd0/eth/fetcher/block_fetcher.go#L682-L686

The broadcast this PR replaced had this same scenario here, that insertChain error doesn't need to(can't) be handled